tests/Doctrine/Tests/Common/Annotations/AnnotationRegistryTest.php

```diff
@@ -65,4 +67,45 @@ protected function getStaticField($class, $field)
 
         return $reflection->getValue();
     }
+
+    public function testStopCallingLoadersIfClassIsNotFound()
+    {
+        AnnotationRegistry::reset();
+        $i = 0;
+        $autoLoader = function($annotation) use (&$i) {
+            $i++;
+            return false;
+        };
+        AnnotationRegistry::registerLoader($autoLoader);
+        AnnotationRegistry::loadAnnotationClass('unloadableClass');
+        AnnotationRegistry::loadAnnotationClass('unloadableClass');
+        AnnotationRegistry::loadAnnotationClass('unloadableClass');
+        $this->assertEquals(1, $i, 'Autoloader should only be called once');
+    }
+
+    public function testStopCallingLoadersAfterClassIsFound()
+    {
+        AnnotationRegistry::reset();
+        $i = 0;
+        $autoLoader = function($annotation) use (&$i) {
+            $i++;
+            return true;
+        };
+        AnnotationRegistry::registerLoader($autoLoader);
+        AnnotationRegistry::loadAnnotationClass(self::class);
+        AnnotationRegistry::loadAnnotationClass(self::class);
+        AnnotationRegistry::loadAnnotationClass(self::class);
+        $this->assertEquals(1, $i, 'Autoloader should only be called once');
+    }
+
+    public function testAddingANewLoaderClearsTheCache()
+    {
+        AnnotationRegistry::reset();
+        AnnotationRegistry::registerLoader('class_exists');
+        AnnotationRegistry::loadAnnotationClass(self::class);
+        AnnotationRegistry::loadAnnotationClass('unloadableClass');
+        AnnotationRegistry::registerLoader('class_exists');
+        self::assertEmpty($this->getStaticField($this->class, 'loaded'));
```

Comment:
This should be checked with an external loader, verifying that it will be called again. Please refrain from reading private state in a test